mlflow/pyspark/ml/__init__.py

```diff
@@ -925,8 +939,56 @@ def _log_posttraining_metadata(estimator, spark_model, params):
         if log_models:
             if _should_log_model(spark_model):
                 # TODO: support model signature
+                from mlflow.models.signature import infer_signature
+
+                def _cast_spark_df_with_vector_to_array(input_spark_df):
+                    """
+                    Finds columns of vector type in a spark dataframe and
+                    casts them to array<double> type.
+
+                    :param input_spark_df:
+                    :return: a spark dataframe with vector columns transformed to array<double> type
+                    """
+                    from functools import reduce
+                    from pyspark.ml.functions import vector_to_array
+                    from pyspark.ml.linalg import VectorUDT
+
+                    vector_type_columns = [
+                        _field.name
+                        for _field in input_spark_df.schema
+                        if isinstance(_field.dataType, VectorUDT)
+                    ]
+                    return reduce(
+                        lambda df, vector_col: df.withColumn(
+                            vector_col, vector_to_array(vector_col)
+                        ),
+                        vector_type_columns,
+                        input_spark_df,
+                    )
+
+                limited_input_df = input_df.limit(INPUT_EXAMPLE_SAMPLE_ROWS)
+
+                def _get_input_example():
+                    return _cast_spark_df_with_vector_to_array(limited_input_df).toPandas()
```

Comment:
Note spark dataframe might includes some column which is unused in spark model training. These columns we are hard to exclude them, no simple API can filter them out.

Reply:
@WeichenXu123 Another great point :). I had to do something similar in the past for automl-toolkit ( cc @BenWilson2) in the past.
To get the true input columns for the training pipeline, we can just step through each stage and extract inputCol. And then do a select on the input_df to get all those columns needed for all the stages in a pipeline. 
But if it is just an estimator (or a pipeline with a single stage), then we can just extract the single features column. 
what do you think of that approach?

Reply:
> it is just an estimator (or a pipeline with a single stage), then we can just extract the single features column.

Seems some model input is not features column, e.g. the ALSModel

> we can just step through each stage and extract inputCol

Some pipeline , follow-on stages might take previous stage output column as input column, we need special handling.

And for some meta-model, e.g. OneVsRestModel, its input column depends on its nested classification model input.

Reply:
I think a safer way is, for pyspark Transformer/Model type, add APIs "getInputColumns" and "getOutputColumns", and implement the API for each specific model. Note many models inherits Predictor Model / Classification Model, which we can implement "getInputColumns" and "getOutputColumns" in their base class.

Reply:
thanks, @WeichenXu123 !
What are your thoughts about this approach (a quick dirty script to showcase the idea):

`
from pyspark.sql import SparkSession, types as t
from pyspark.ml.feature import IndexToString, StringIndexer, VectorAssembler
from pyspark.ml import Pipeline
from pyspark.ml.classification import LogisticRegression
import re
from pyspark.sql.utils import IllegalArgumentException

spark = SparkSession.builder.getOrCreate()

training_df = spark.createDataFrame(
    [
        (0, "a", "a", "b"),
        (1, "b", "a", "b"),
        (2, "c", "a", "b"),
        (3, "a", "a", "b"),
        (4, "a", "a", "b"),
        (5, "c", "a", "b"),
    ]
).toDF("id", "category", "not_a_feature_1", "not_a_feature_2")

string_indexer = StringIndexer(inputCol="category", outputCol="label").fit(training_df)
pipeline = Pipeline(
    stages=[
        string_indexer,
        VectorAssembler(inputCols=["id"], outputCol="features"),
        LogisticRegression(),
        IndexToString(
            inputCol="prediction", outputCol="originalLabel", labels=string_indexer.labels
        ),
    ]
)

pipeline_model = pipeline.fit(training_df)
params = pipeline_model


def do_pipeline_transform(spark_df):
    return pipeline_model.transform(spark_df)


def get_struct_type_by_cols(cols):
    if len(cols) > 0:
        return t.StructType([_field for _field in training_df.schema.fields if _field.name in cols])
    return []


def get_feature_cols(input_fields=None):
    if input_fields is None:
        input_fields = set({})
        df = spark.createDataFrame([()])
    else:
        df = spark.createDataFrame(
            spark.sparkContext.emptyRDD(), get_struct_type_by_cols(input_fields)
        )
    try:
        do_pipeline_transform(df)
    except IllegalArgumentException as iae:
        col_name_search = re.search("(.*) does not exist.", iae.desc, re.IGNORECASE)
        if col_name_search:
            col_name = col_name_search.group(1)
            input_fields.add(col_name)
            get_feature_cols(input_fields)
    return input_fields


required_cols = get_feature_cols()

print(
    t.StructType(
        [_field for _field in training_df.schema.fields if _field.name in required_cols]
    ).json()
)
`
Output:
`
{"fields":[{"metadata":{},"name":"id","nullable":true,"type":"long"}],"type":"struct"}
`

We can add unit tests around that message that spark pipeline/transformer internals throw to make sure it doesn't change or keep up with it. 
Thoughts?

Reply:
Very interesting idea! Are there any significant performance implications? One drawback is that this doesn't solve the "useless columns" issue identified by @WeichenXu123: just because `do_pipeline_transform()` fails for a set of input columns doesn't mean that all of those input columns are required; it just means that one or more required columns, which may be explored later, are missing from the set.

Reply:
> I think a safer way is, for pyspark Transformer/Model type, add APIs "getInputColumns" and "getOutputColumns", and implement the API for each specific model. Note many models inherits Predictor Model / Classification Model, which we can implement "getInputColumns" and "getOutputColumns" in their base class.

@WeichenXu123 Is the implementation going to be standard for `Predictor Model` / `Classification Model`? If so, can we implement the same logic for these models in MLflow right now and contribute it to PySpark ML after? It's okay to begin with a limited set of models.

Reply:
@dbczumar 
that last print statement above does that check to filter out columns that aren't needed for any of the pipelines' stages. I just edited that line above to make it part of required_cols. thoughts?

Reply:
Also, I am doing some local testing to see if there could be performance issues with this approach, which is one of the reasons I am using an empty DataFrame in the example above with the right schema,  but the way all the internal spark transformers are written, they [validate schema](https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/ml/Pipeline.scala#L134) before any fit/transform:

Reply:
Option 2 is to do something like this:

```
from pyspark.sql import SparkSession, types as t
from pyspark.ml.feature import IndexToString, StringIndexer, VectorAssembler
from pyspark.ml import Pipeline
from pyspark.ml.classification import LogisticRegression
spark = SparkSession.builder.getOrCreate()

training_df = spark.createDataFrame(
    [
        (0, "a", "b", "c", "d", 1.0),
        (1, "a", "b", "c", "d", 0.0),
        (2, "a", "b", "c", "d", 2.0),
        (3, "a", "b", "c", "d", 3.0),
        (4, "a", "b", "c", "d", 1.0),
        (5, "a", "b", "c", "d", 1.0),
    ]
).toDF("id", "feature_a", "feature_b", "not_a_feature_1", "not_a_feature_2", "label")

pipeline = Pipeline(
    stages=[
        StringIndexer(inputCols=["feature_a", "feature_b"], outputCols=["feature_a_indexed", "feature_b_indexed"]),
        VectorAssembler(inputCols=["id", "feature_a_indexed", "feature_b_indexed"], outputCol="features"),
        LogisticRegression()
    ]
)

pipeline_model = pipeline.fit(training_df)

pipeline_input_cols = set({})
label_col = None

pipeline_model.transform(training_df.drop("categories")).show()

for stage in pipeline_model.stages:
    if stage.hasParam("inputCol"):
        pipeline_input_cols.add(stage.extractParamMap().get(stage.inputCol))
    if stage.hasParam("inputCols"):
        input_cols = stage.extractParamMap().get(stage.inputCols)
        if input_cols is not None:
            pipeline_input_cols.update(input_cols)
    if stage.hasParam("featuresCol"):
        pipeline_input_cols.add(stage.extractParamMap().get(stage.featuresCol))
    if stage.hasParam("labelCol"):
        pipeline_input_cols.add(stage.extractParamMap().get(stage.featuresCol))

required_cols = set(training_df.columns).intersection(pipeline_input_cols)
print(required_cols)
```
output:
```
{'feature_a', 'feature_b', 'id'}
```

Checking [here](https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/ml/param/shared/sharedParams.scala) in the shared params, we could have: `inputCol`, `inputCols`, `labelCol`, `outputCol`, `outputCols`, `featuresCol`. We should only need to look at `inputCol`,  `inputCols` to be the input columns needed for pipeline stages, but ALS seems like an exception with `itemCol` and `userCol`. 

The first option example above will cover all stages (subject to further unit tests) and the second option just needs extra steps to cover an outlier pipeline stage like ALS (to handle custom params other than `inputCol`/`inputCols` that refer to the input columns from the training df).

Reply:
Could you add a presence check for 'itemCol' and 'userCol' (both must be present) for ALS to the logic you have in Option 2? 
Also, nit: `pipeline_input_cols = set()`

Reply:
> @dbczumar that last print statement above does that check to filter out columns that aren't needed for any of the pipelines' stages. I just edited that line above to make it part of input_required_struct. thoughts?

Ah, my mistake. I missed the `re.search("(.*) does not exist.", iae.desc, re.IGNORECASE)` condition. Seems like this will work. @WeichenXu123 thoughts?

Reply:
I incline to use this approach https://github.com/mlflow/mlflow/pull/5719#discussion_r854314443
which looks like a clever way.

One concern is, some model might raise error when hit empty input dataframe. So we should wrap try ... catch around it and if it failed, we log warning and skip logging signature.

Reply:
> > I think a safer way is, for pyspark Transformer/Model type, add APIs "getInputColumns" and "getOutputColumns", and implement the API for each specific model. Note many models inherits Predictor Model / Classification Model, which we can implement "getInputColumns" and "getOutputColumns" in their base class.
> 
> @WeichenXu123 Is the implementation going to be standard for all subclasses of `Predictor Model` / `Classification Model`? If so, can we implement the same logic for these models in MLflow right now and contribute it to PySpark ML after? It's okay to begin with a limited set of models.

"implement the same logic for these models in MLflow side", we might need to implement many class including many transformers. Currently I incline to use https://github.com/mlflow/mlflow/pull/5719#discussion_r854314443 approach but eventually we should contribute to pyspark ML and implementing the safe getInputColumns/getOutputColumns logic